### PR TITLE
cudaPackages.autoAddX: Deprecate

### PR DESCRIPTION
**Pull request by jonringer**
## Description of changes

Draft until the following have been merged:
-  #300277
-  #300278
-  #300287
-  #300288
-  #300290
-  #300292
-  #300293
-  #300294
-  #300295
-  #300296
-  #300297
-  #300298
-  #300299
-  #300300
-  #300301
-  #300302

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/cuda-modules/setup-hooks/extension.nix [hunk 1]
```diff
@@ -1,10 +1,5 @@
 final: _: {
-  # TODO:
```
Comment (SomeoneSerge) [inline pkgs/development/cuda-modules/setup-hooks/extension.nix:2]: Don't we want to deprecate `cudaPackages.autoFixElfFiles` too?

Comment (jonringer) [reply pkgs/development/cuda-modules/setup-hooks/extension.nix:2]: Forgot it was part of the refactor. I'll include it

Comment (jonringer) [reply pkgs/development/cuda-modules/setup-hooks/extension.nix:2]: done

**Review by ConnorBaker — APPROVED** [review]

**Comment by ConnorBaker** [comment]
@jonringer looks good to me -- feel free to merge after `ofborg-eval` in CI finishes!

File: pkgs/development/cuda-modules/aliases.nix [hunk 2]
```diff
@@ -1,4 +1,20 @@
 # Packges which have been deprecated or removed from cudaPackages
-final: prev: {
+final: prev:
+let
+  mkDeprecation =
+    oldName: newName: newPkg:
+    final.lib.warn "cudaPackages.${oldName} is deprecated, use ${newName} instead" newPkg;
```
Comment (SomeoneSerge) [inline pkgs/development/cuda-modules/aliases.nix:6]: `mkRenamed`/`mkRenamedAttribute` (analogous to `mkRenamedOptionModule`) is probably clearer, especially if we later want to deprecate things (e.g. `cudatoolkit`) by actually removing them

Comment (jonringer) [reply pkgs/development/cuda-modules/aliases.nix:6]: Sure, sounds good with me

**Comment by ConnorBaker** [comment]
The formatter is displeased now?

**Comment by jonringer** [comment]
Yea.... need alter my vim setup to now run nixfmt on stuff

**Review by SomeoneSerge — APPROVED** [review]

**Comment by ConnorBaker** [comment]
Merging since CI passed. Thank you @jonringer :)

**Comment by jonringer** [comment]
After 2.5 years, finally checked this off of my list of "weirdness budge"

**Pull request merged**